### PR TITLE
Reservoir and stream reach intersection

### LINKED ISSUE #31: ID reaches that are reservoirs
We want to identify which stream reaches in the DRB are actually reservoirs. I think the process should be:

1. Use a nationally-available reservoir or lake database, like the [NHD](https://nhd.usgs.gov/userGuide/Robohelpfiles/NHD_User_Guide/Feature_Catalog/Hydrography_Dataset/NHDPoint/Reservoir.htm) (it does specify reservoir and reservoir type) or [GRanD](https://sedac.ciesin.columbia.edu/data/set/grand-v1-dams-rev01) to get polygons. For now, I like the idea of GRanD because it is reservoir centric, and has volumes and other metadata associated with it. It was intended to capture reservoirs of >0.1km3, but notes there are smaller reservoirs when data were available.

2. Subset the reservoirs to those within our DRB footprint. Maintain these polygons in the pipeline - will be nice for plotting/mapping purposes.

2. Find all of the reaches in [our modeling network](https://github.com/USGS-R/delaware-model-prep/blob/master/1_network.yml#L45) that are within the polygon.

3. Create a dataframe of all network reaches with a field indicating whether the reach is a reservoir, and maintain some reservoir ID.

4. Create a dataframe of all reservoirs in the network that maintains the unique identifier and metadata (e.g., surface area, volume).

### PR DESCRIPTION
**Pull request by wdwatkins**
Fixes #31 

This creates the target `1_network/out/subseg_reservoir_mapping.rds.ind`, a data frame of all reach subsegments with all associated information from the GRaND dam/reservoir files, as well as some computed distances between the dam, reservoir centroid, and outlet/inlet and total percent of each subsegment that overlaps with a reservoir.  

I included a plotting function to make plots of each reservoir and its overlapping reaches.  It is not built into the pipeline.

This should be fully generalizable to a different basin.  It might work for all CONUS, this only takes a minute or two to run, but I imagine there would be some more edge cases to deal with.

### CODE REVIEW

File: 1_network.yml [hunk 1]
```diff
@@ -31,6 +32,33 @@ targets:
...
+  #download reservoirs and dams from drive
+  #gd_confirm targets are only used to intialize indicator files, should be left commented
+  #1_network/in/dams-rev01-global-shp.zip.ind:
+  #  command: gd_confirm_posted(target_name)
+  1_network/in/dams-rev01-global-shp.zip:
+    command: gd_get('1_network/in/dams-rev01-global-shp.zip.ind')
+  1_network/out/global_dams.rds.ind:
+    command: zipped_shp_to_rds(out_ind = target_name,
+      zip = '1_network/in/dams-rev01-global-shp.zip', 
+      extract_dir = I('1_network/in/dams-rev01-global-shp'))
```
Comment (jread-usgs) [inline 1_network.yml:45]: cannot resist urge to comment on PR.....halp

Is the `extract_dir` needed here? If you are not planning on referencing the unzipped shape file files in any way in the future, do you want to throw them away instead? Otherwise, they seem to take up unnecessary local space. What if you unzip to a temp dir (and that way you don't need the additional argument) and/or unzip to temp and then `unlink` after reading as `sf`?

Comment (wdwatkins) [reply 1_network.yml:45]: I didn't want to unzip directly into `tmpdir`, since then it would be clunky to delete the contents without blowing away everything else in `tmpdir` that is created for the R session.  But I could just create these same folders in `tmpdir` using the exdir argument.  I suppose it doesn't need to be exposed to the user.

File: 1_network/src/network_dam_res_reaches.R [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+intersect_network_with_reservoirs <- function(stream_network_file, dams_reservoirs_file, out_ind) {
+  
+  stream_network <- readRDS(stream_network_file)
+  dams_reservoirs <- readRDS(dams_reservoirs_file)
+  
+  #create some non-sf tibbles to allow joining by ID rather than geometry
+  network_edges_tibble <- as_tibble(stream_network$edges)
+  network_vertices_tibble <- as_tibble(stream_network$vertices)
+  network_vertices_tibble_separated <- network_vertices_tibble %>% 
+    tidyr::separate_rows(ends_subseg, sep = ";")
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply 1_network/src/network_dam_res_reaches.R:51]: Yeah, it was new to me too!  I just ran across it in the manual pages by accident

File: 1_network.yml (hunk 1 above)
Comment (jread-usgs) [reply 1_network.yml:45]: yeah, you can also create a folder within the tempdir for the unzips to go into and then unlink that whole file, like
```r
unz_dir <- file.path(tempdir(),'shapefile')
dir.create(unz_dir)
# do your thing
unlink(unz_dir)
```

File: 1_network/src/network_dam_res_reaches.R [hunk 3]
```diff
@@ -2,10 +2,11 @@
 #' @param extract_dir char directory to unzip to, and read shapefile from
 #' @param out_ind char indicator file to represent output
 zipped_shp_to_rds <- function(zip, out_ind) {
-  extract_dir <- file.path(tempdir(), 'shapefile_tmp')
+  extract_dir <- file.path(tempdir(), 
+                           tools::file_path_sans_ext(basename(zip)))
   unzip(zip, exdir = extract_dir)
   read_sf(extract_dir) %>% saveRDS(file = as_data_file(out_ind))
-  unlink(extract_dir)
+  unlink(extract_dir, recursive = TRUE)
```
Comment (wdwatkins) [inline 1_network/src/network_dam_res_reaches.R:9]: Unlink just silently does nothing if you pass it a directory name without `recursive=TRUE` 😬

File: 1_network/src/network_dam_res_reaches.R [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+#' @param dams_file directory of dams shapefile
+#' @param reservoirs_file directory of reservoir shapefile
+#' @param boundary polygon of basin boundaries to subset dams/reservoirs by
+#' @param outind output indicator file
+filter_dams_reservoirs_by_boundary <- function(dams_shp_ind, reservoirs_shp_ind, 
+                                               boundary_ind, out_ind) {
+  #dams are points, reservoirs are polygons
+  dams_shp <- readRDS(sc_retrieve(dams_shp_ind)) %>% 
+    st_transform(crs = 102039) %>%
```
[the remark this answers is not in the document]
Comment (wdwatkins) [reply 1_network/src/network_dam_res_reaches.R]: Ok, Googling this a bit more and looking at the second warning, I think the EPSG code 102039 might be specific to ESRI or something?  I see the projection referenced as `EPSG:5070` in some other places.  Maybe whatever library defines EPSG codes is different on your machine, and does not include 102039?  It is what `st_crs` returns for the boundary shapefile though:
image

What does that return for you? 

In any case, I think using the CRS from the boundary shapefile rather than hardcoding it is probably the right choice.  You might end up with fighting indicator files though if the boundary shapefile is read differently though 😕

[the remark this answers is not in the document]
Comment (limnoliver) [reply 1_network/src/network_dam_res_reaches.R]: Mine returns the same `102039`. What do you mean if the boundary shapefile is read differently? You mean if it returns a different ESPG?

[the remark this answers is not in the document]
Comment (wdwatkins) [reply 1_network/src/network_dam_res_reaches.R]: Well, your second warning seems to say that GDAL (which I assume `sf` is using under the hood here) doesn't have the 102039 projection defined.  This was not an issue for me, so I must have a different GDAL installed.  I guess the error happened in st_transform though instead of when the file is read.  I had thought maybe your machine would substitute a different projection when reading it.

Thinking about this a bit more, I bet that using st_transform(EPSG code) depends on GDAL having that EPSG code defined (yours apparently doesn't).  But if you use `st_transform(crs = sf::st_crs(boundary))` you are just using the projection definition stored in the boundary `object`, so GDAL having the projection doesn't matter.   Maybe there is a subtle difference in my GDAL's definition of the 102039 projection compared to what the `boundary` object uses, that would cause the object hashes to change with your PR?

TL;DR here is I think this is good to go with your change 😄

[the remark this answers is not in the document]
Comment (limnoliver) [reply 1_network/src/network_dam_res_reaches.R]: Yes, you are correct that it happened in the transform -- I could read the file in just fine. 

Thanks for this follow-up, and for the note about potential changes in hashes and indicator files. That is giving me a clue to mystery rebuilds that we have in the pipeline.

**Comment by wdwatkins** [comment]
Go ahead and merge if you are happy, I don't have anything else to add
